rule_builder/options.py

```diff
@@ -0,0 +1,90 @@
+import dataclasses
+import importlib
+import operator
+from collections.abc import Callable, Iterable
+from typing import Any, Generic, Literal, Self, cast
+
+from typing_extensions import TypeVar, override
+
+from Options import CommonOptions, Option
+
+Operator = Literal["eq", "ne", "gt", "lt", "ge", "le", "contains"]
+
+OPERATORS: dict[Operator, Callable[..., bool]] = {
+    "eq": operator.eq,
+    "ne": operator.ne,
+    "gt": operator.gt,
+    "lt": operator.lt,
+    "ge": operator.ge,
+    "le": operator.le,
+    "contains": operator.contains,
+}
+operator_strings: dict[Operator, str] = {
+    "eq": "==",
+    "ne": "!=",
+    "gt": ">",
+    "lt": "<",
+    "ge": ">=",
+    "le": "<=",
+}
+
+T = TypeVar("T")
+
+
+@dataclasses.dataclass(frozen=True)
+class OptionFilter(Generic[T]):
+    option: type[Option[T]]
+    value: T
+    operator: Operator = "eq"
+
+    def to_dict(self) -> dict[str, Any]:
+        """Returns a JSON compatible dict representation of this option filter"""
+        return {
+            "option": f"{self.option.__module__}.{self.option.__name__}",
+            "value": self.value,
+            "operator": self.operator,
+        }
+
+    def check(self, options: CommonOptions) -> bool:
+        """Tests the given options dataclass to see if it passes this option filter"""
+        option_name = next(
+            (name for name, cls in options.__class__.type_hints.items() if cls is self.option),
+            None,
+        )
+        if option_name is None:
+            raise ValueError(f"Cannot find option {self.option.__name__} in options class {options.__class__.__name__}")
+        opt = cast(Option[Any] | None, getattr(options, option_name, None))
+        if opt is None:
+            raise ValueError(f"Invalid option: {option_name}")
+
+        return OPERATORS[self.operator](opt.value, self.value)
```

Comment:
Hmmm. Are you sure about using opt.value instead of opt? We have a lot of fancy dunder overrides for option classes that allow for fancier comparisons. But I'm also not immediately sure if I can claim that those dunder overrides are "complete". I have not had to actually get option.value in a long time, but there might be one or two niche things we have forgotten.

Reply:
My main concern is that people might expect string comparisons to work, but it's not the worst thing in the world if they don't.